File: leetcode/0364-nested-list-weight-sum-ii/solution.py
from typing import List


# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
# class NestedInteger:
#     def isInteger(self) -> bool:
#         """
#         @return True if this NestedInteger holds a single integer, rather than a nested list.
#         """
#
#     def getInteger(self) -> int:
#         """
#         @return the single integer that this NestedInteger holds, if it holds a single integer
#         Return None if this NestedInteger holds a nested list
#         """
#
#     def getList(self) -> [NestedInteger]:
#         """
#         @return the nested list that this NestedInteger holds, if it holds a nested list
#         Return None if this NestedInteger holds a single integer
#         """


class Solution:
    def depthSumInverse(self, nestedList: List['NestedInteger']) -> int:
        def max_depth(items, depth):
            deepest = depth
            for item in items:
                if not item.isInteger():
                    deepest = max(deepest, max_depth(item.getList(), depth + 1))
            return deepest

        bottom = max_depth(nestedList, 1)

        def helper(items, depth):
            total = 0
            for item in items:
                if item.isInteger():
                    total += item.getInteger() * (bottom - depth + 1)
                else:
                    total += helper(item.getList(), depth + 1)
            return total

        return helper(nestedList, 1)

File: leetcode/0364-nested-list-weight-sum-ii/Solution.java
import java.util.List;

/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * public interface NestedInteger {
 *
 *     // @return true if this NestedInteger holds a single integer, rather than a nested list.
 *     public boolean isInteger();
 *
 *     // @return the single integer that this NestedInteger holds, if it holds a single integer
 *     // Return null if this NestedInteger holds a nested list
 *     public Integer getInteger();
 *
 *     // @return the nested list that this NestedInteger holds, if it holds a nested list
 *     // Return null if this NestedInteger holds a single integer
 *     public List<NestedInteger> getList();
 * }
 */
class Solution {
    public int depthSumInverse(List<NestedInteger> nestedList) {
        int bottom = maxDepth(nestedList, 1);
        return helper(nestedList, 1, bottom);
    }

    private int maxDepth(List<NestedInteger> items, int depth) {
        int deepest = depth;
        for (NestedInteger item : items) {
            if (!item.isInteger()) {
                deepest = Math.max(deepest, maxDepth(item.getList(), depth + 1));
            }
        }
        return deepest;
    }

    private int helper(List<NestedInteger> items, int depth, int bottom) {
        int total = 0;
        for (NestedInteger item : items) {
            if (item.isInteger()) {
                total += item.getInteger() * (bottom - depth + 1);
            } else {
                total += helper(item.getList(), depth + 1, bottom);
            }
        }
        return total;
    }
}
